### PR TITLE
Refactor streaming code to support interop testing

### LINKED ISSUE #16119: Add integration test with mock kubelet 1.0 backend to verify exec subprotocol fallback
PR #13322 fixed a variety of bugs with exec/attach, but unfortunately introduced some backwards incompatibility when a newer kubectl that included the fixes attempted to exec against a 1.0.x kubelet.

#15961 adds support for subprotocol negotiation for exec, attach, and portfoward, and restores the ability for the latest kubectl to talk to 1.0.x kubelets for exec.

We need a way to test that the restored behavior (new kubectl, 1.0 kubelet) is working.

cc @smarterclayton

### PR DESCRIPTION
**Pull request by ncdc**
Refactor exec/attach/port forward client and server code to better
support interop testing of different client and server subprotocol
versions.

Fixes #16119

### CODE REVIEW

**Comment by ncdc** [comment]
@smarterclayton @lavalamp @liggitt here's my WIP on #16119

**Comment by ncdc** [comment]
@smarterclayton @lavalamp are you interested in having me continue with this? If so, I'll rebase and would appreciate feedback. If not, we can close it.

**Comment by smarterclayton** [comment]
I think it's a good thing to do, it's your call though whether it's more
important than other changes you have.

**Comment by ncdc** [comment]
I'm looking for feedback on the current state of the PR to see if it's in the right direction. Also, when we move to HTTP/2, it's highly likely that the `httpstream` package's `Connection` and `Stream` types will become deprecated; they're not used for the websocket implementation, and the HTTP/2 implementation pattern will likely be operating solely on the request and response bodies and muxing stdout/stderr as needed (similar to how you did it in the wsstream).

**Comment by ncdc** [comment]
But I do also think helping David out with the group / version / kind apiserver changes are more important.

**Comment by smarterclayton** [comment]
The structure looks fine to me.  Anything is better than the mess before.  This is less bad.

**Comment by ncdc** [comment]
@lavalamp @smarterclayton @liggitt PR updated (finally), PTAL

**Comment by ncdc** [comment]
I see I have some import direction issues that are causing the `httptest.serve` flag to leak into most/all of the CLI commands. Working on a fix now.

**Comment by ncdc** [comment]
Apparently the leaking is occurring because `pkg/kubectl/cmd/exec.go` and `pkg/kubectl/cmd/attach.go` are importing `pkg/kubelet/server` to make use of my new `server.SupportedStreamingProtocols` var in this PR. I'm not sure why importing that package would yield an `httptest` flag coming into the picture, though. It looks like the `server` package only imports `httptest` in `server_test.go`...

**Comment by ncdc** [comment]
Well this was fun: pkg/kubelet/server -> pkg/kubelet/container -> https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/container/runtime_mock.go#L22 which ultimately imports httptest

**Comment by ncdc** [comment]
Ok, so the question is this: is it appropriate to put `SupportedStreamingProtocols` in `pkg/kubelet/server`? If so, then we are going to end up with `--httptest.serve` being a new no-op flag in all `kubectl` commands until we can resolve what to do about `pkg/kubelet/container/runtime_mock.go`, since that's ultimately responsible for pulling in httptest.

**Comment by yujuhong** [comment]
@ncdc, if the `runtime_mock.go` file is causing problems, I can move it to the test file that's using the mock until we have better solutions.

**Comment by smarterclayton** [comment]
I'd prefer to fix runtime_mock.go if at all possible (by changing how the flag is registered).

**Comment by ncdc** [comment]
@smarterclayton any regular (non-test) file that pulls in testify's mock package will end up transitively pulling in net/http/httptest, and the httptest package has a flag that is registered as soon as you import the package. It's not just the Kubernetes code that has an issue... I found some code that the Mesos cloud provider imports (another Mesos library) that also has a mock, so even if we fix it here, it's not a complete fix...

**Comment by smarterclayton** [comment]
Can we blacklist it?

**Comment by yujuhong** [comment]
FYI, there is a new issue opened to get rid of the test flags: #21114
@timstclair proposed using a `test` build tag.

**Comment by ncdc** [comment]
FYI I plan on picking this back up after 1.2 is branched

**Comment by ncdc** [comment]
Rebased and refactored a bit

**Comment by ncdc** [comment]
@kubernetes/rh-cluster-infra PTAL if you have some time to review

File: pkg/kubelet/server/remotecommand/websocket.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	"time"
+
+	"k8s.io/kubernetes/pkg/httplog"
+	"k8s.io/kubernetes/pkg/util/wsstream"
+
+	"github.com/golang/glog"
+)
+
+// standardShellChannels returns the standard channel types for a shell connection (STDIN 0, STDOUT 1, STDERR 2)
+// along with the approprxate duplex value
```
Comment (smarterclayton) [inline pkg/kubelet/server/remotecommand/websocket.go]: Describe what protocols this supports.  Typo in approximate

File: pkg/kubelet/server/remotecommand/httpstream.go [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+	defer close(stop)
+WaitForStreams:
+	for {
+		select {
+		case stream := <-streams:
+			streamType := stream.Headers().Get(api.StreamType)
+			switch streamType {
+			case api.StreamTypeError:
+				ctx.errorStream = stream
+				// this defer statement shouldn't be here - it should be in
```
Comment (smarterclayton) [inline pkg/kubelet/server/remotecommand/httpstream.go]: Make this a TODO

Comment (smarterclayton) [inline pkg/kubelet/server/remotecommand/httpstream.go]: Or change the comment to imply that v2 fixes this and so it's not a TODO, but a back compat issue.

File: pkg/kubelet/server/remotecommand/httpstream.go [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+type options struct {
+	stdin           bool
+	stdout          bool
+	stderr          bool
+	tty             bool
+	expectedStreams int
+}
+
+// newRemoteCommandOptions creates a new options from the Request.
+func newOptions(req *http.Request) (*options, error) {
```
Comment (smarterclayton) [inline pkg/kubelet/server/remotecommand/httpstream.go]: mismatch

**Comment by smarterclayton** [comment]
Are the existing E2E tests sufficient to cover this function?

**Comment by ncdc** [comment]
> Are the existing E2E tests sufficient to cover this function?

They cover whatever the current protocol is. The new unit tests cover version skew testing.

**Comment by smarterclayton** [comment]
Changes LGTM

**Comment by smarterclayton** [comment]
Squash and I'll tag

**Comment by ncdc** [comment]
Squashed

**Comment by ncdc** [comment]
Failure was #22639 / #23607

**Pull request merged**